### PR TITLE
Enabled UVC (USB Video Class) camera and pre-recorded H.264 video file as input source.

### PR DESCRIPTION
**Pull request by Devinwong**
Enabled UVC (USB Video Class) camera and pre-recorded H.264 video file as input source.
Only enabled in object detection as an example.

### CODE REVIEW

**Comment by Devinwong** [comment]
Please don't merge until we fix the base image acr credentials issue

File: azureeyemodule/README.md
Comment (MaxStrange) [inline azureeyemodule/README.md:867]: Typo. `/video:/video` should be `video:/video`, right?

File: azureeyemodule/app/model/objectdetector.cpp
Comment (MaxStrange) [inline azureeyemodule/app/model/objectdetector.cpp:99]: While you are fixing other typos, you might as well fix this one too: `These are th enodes` -> `These are the nodes`. Thanks

File: azureeyemodule/app/model/ssd.cpp
Comment (MaxStrange) [inline azureeyemodule/app/model/ssd.cpp:219]: Remove extra space to align the curly braces with the if's indentation level.

File: azureeyemodule/app/model/ssd.cpp
Comment (MaxStrange) [inline azureeyemodule/app/model/ssd.cpp:49]: Double check that we don't already have a function that does this in the helper module (util/helper.[c/h]pp). If we don't, please move this function to that file, so other modules can use it.

File: azureeyemodule/app/model/ssd.cpp
Comment (MaxStrange) [inline azureeyemodule/app/model/ssd.cpp:53]: Place the curly braces on their own line in this code base. Thaaaanks

File: azureeyemodule/app/model/objectdetector.cpp
Comment (MaxStrange) [inline azureeyemodule/app/model/objectdetector.cpp:93]: This seems like a lot of code duplication. Are you sure you can't combine this method with the normal pull_data method? Don't force it if you think it will result in uglier/harder-to-maintain code though.

File: azureeyemodule/app/model/objectdetector.hpp
Comment (MaxStrange) [inline azureeyemodule/app/model/objectdetector.hpp:31]: Doesn't the user pass in the "video:" as part of the arg on the command line?

File: azureeyemodule/app/model/ssd.cpp
Comment (MaxStrange) [inline azureeyemodule/app/model/ssd.cpp:15]: Interesting. We have previously been rendering the inferences using traditional OpenCV library calls, rather than incorporating the rendering into the G-API.

I don't care which one we do, but we should do it the same way for all of the models. That being the case, it is way easier to just keep doing the way we've been doing it. Is there a reason to change to using G-API for rendering? I'm guessing it is not as easy or intuitive but may be more efficient.

File: azureeyemodule/app/model/ssd.cpp
Comment (MaxStrange) [inline azureeyemodule/app/model/ssd.cpp:92]: Two things:
1. Please put parentheses around the different conditionals in this if statement, just because it is easier to read and better practice (since C often has surprising operator precedence rules).
2. This whole if statement is the type of thing I would love to avoid by combining the `compile_cv_graph_uvc_video` with `compile_cv_graph`. I feel like it should be not too hard, since the only thing that should really change is the input source. Then you could also combine `pull_data_uvc_video` with `pull_data`.

File: azureeyemodule/app/model/ssd.cpp
Comment (MaxStrange) [inline azureeyemodule/app/model/ssd.cpp:180]: Again, this is a ton of code duplication. It would be better if we could combine this method with the other `compile_graph` method, or at the very least, use the one in the super class.

The only reason for this method to be different from the one in the super class (that I can tell) is that you are using the G-API to do the inference rendering (bounding box drawing). That's another reason in favor of not doing it that way, and instead just drawing the bounding boxes like we've been doing it, using the traditional OpenCV library calls.

**Review by MaxStrange — CHANGES_REQUESTED** [review]
Great job, Devin for enabling this feature! A few typos here and there, and one big change that I would prefer: could you merge the new `compile` and `pull` methods you created with the ones that we already have?

This would cut down on code duplication dramatically and make the logic a little simpler.

As far as I can tell, the reason you need to break out a new `pull` and `compile` method is that you are now using the G-API to do the bounding box drawing. Unless there's a super good reason to go that route (in which case we should do it for all the models), I think we should keep doing it the way we've been doing it (by calling OpenCV methods instead). I am guessing it is more flexible, easier to reason about, and easier to write, but probably not as efficient as doing it directly in the G-API.

Does that all make sense?

File: azureeyemodule/README.md
Comment (Devinwong) [reply azureeyemodule/README.md:867]: Actually my video folder is located in root. That's why I wrote /video as the first one. So /video:/video is correct in my opinion.
Or did I miss anything?

File: azureeyemodule/app/model/objectdetector.cpp
Comment (Devinwong) [reply azureeyemodule/app/model/objectdetector.cpp:99]: good catch. thanks

File: azureeyemodule/app/model/ssd.cpp
Comment (Devinwong) [reply azureeyemodule/app/model/ssd.cpp:49]: there is an existing one. use that instead

File: azureeyemodule/app/model/objectdetector.hpp
Comment (Devinwong) [reply azureeyemodule/app/model/objectdetector.hpp:31]: Yes. so I am using this prefix to parse the file path behind this keyword. Does this answer your question?

File: azureeyemodule/app/model/objectdetector.cpp
Comment (Devinwong) [reply azureeyemodule/app/model/objectdetector.cpp:93]: image
To give you more context, the flow didn't change a lot from the default one to this uvc_video one but the elements of cv::gout() did. One reason is because for UVC and video, G-API doesn't generate any h264 outputs. So I prefer to keep it as current implementation. What do you think?

File: azureeyemodule/app/model/ssd.cpp
Comment (Devinwong) [reply azureeyemodule/app/model/ssd.cpp:15]: This is a long story. Let me try my best to make it clear here.

If you take a look at the custom BBoxes in ssd.cpp, namely, G_API_OP(BBoxes, <GPrims(GDetections)>, "sample.custom.b-boxes"), it's actually empty in the function body. So actually it's not rendering in G-API, but still with the traditional OpenCV library. Therefore, in the compile_cv_graph_uvc_video in objectdetection.cpp, I still call handle_bgr_output(). 

But why I used the rendering? It's because from the latest Intel drop 016 sample, they suggested to use such implementation to get the raw frame from a UVC camera or video file. On the other hand, in the original compile_cv_graph(), we desync'd the img_ts from img, which is
```
cv::GMat img = cv::gapi::copy(cv::gapi::streaming::desync(preproc));
cv::GOpaque<int64_t> img_ts = cv::gapi::streaming::timestamp(img);
```

but it's not needed when using uvc camera and video file.

File: azureeyemodule/app/model/ssd.cpp
Comment (Devinwong) [reply azureeyemodule/app/model/ssd.cpp:92]: 1. done
2. There are quite a lot differences between 2 functions, for example, img and img_ts desync, kernel containing custom G-API op, input source. That's why I made them into 2 separate functions. Any suggestions?

File: azureeyemodule/app/model/ssd.cpp
Comment (Devinwong) [reply azureeyemodule/app/model/ssd.cpp:180]: as answered above.

File: azureeyemodule/README.md
Comment (MaxStrange) [reply azureeyemodule/README.md:867]: You say `--input=video:/video/blah` but then in the description you say `/video:/video`. If you don't think it is a typo, then I am misunderstanding it, and it doesn't really matter. Close this comment out or whatever.

File: azureeyemodule/app/model/objectdetector.hpp
Comment (MaxStrange) [reply azureeyemodule/app/model/objectdetector.hpp:31]: I don't know. I just figured if the user was already writing this, why do we need it here. But also, if it works, I'm not going to worry about it.

File: azureeyemodule/app/model/ssd.cpp
Comment (MaxStrange) [reply azureeyemodule/app/model/ssd.cpp:15]: Hmm. Okay. Yes, let's make sure that we have a uniform way of doing it.

My preferences in order of what I would be most happy with, to what I would grumble about:

1. Have all object detectors use a common super-class's G-API graph and associated methods, regardless of input source. It sounds like this would require a few things: 1) dealing with the fact that H.264 is not output from anything but the MIPI camera (seems like we could get around that with an if statement) and 2) dealing with the timestamps; I think we will need the timestamps for the graphs when USB camera is the input, since we will still have the same problem with slow networks that we have with the MIPI - that the inferences get out of sync with the camera frames. We may also need to incorporate the inference thingy into the pipeline, but it is still not clear to me why.
2. Have all object detectors split their graphs into the MIPI one and the USB/file one, but they all use the same ones in the super class.
3. Have all models have a completely unique version of USB/file graph. Sad face.

I'd like to resolve this before merging in.

File: azureeyemodule/README.md
Comment (Devinwong) [reply azureeyemodule/README.md:867]: the _video:_ in --input is a static prefix, while the _/video_ in -v is a folder.
Maybe next time I should change the folder name to other name to better distinguish.
It's not a typo here.

File: azureeyemodule/app/model/ssd.cpp
Comment (Devinwong) [reply azureeyemodule/app/model/ssd.cpp:15]: After further discussions, we will keep the current PR as is

**Review by MaxStrange — APPROVED** [review]

**Pull request merged**
